bigquery/google/cloud/bigquery/dbapi/cursor.py

```diff
@@ -0,0 +1,280 @@
+        """
+        total_rows = 0
+        num_dml_affected_rows = query_results.num_dml_affected_rows
+
+        if (query_results.total_rows is not None
+                and query_results.total_rows > 0):
+            total_rows = query_results.total_rows
+        if num_dml_affected_rows is not None and num_dml_affected_rows > 0:
+            total_rows = num_dml_affected_rows
+        self.rowcount = total_rows
+
+    def _format_operation_list(self, operation, parameters):
+        """Formats parameters in operation in the way BigQuery expects.
+
+        The input operation will be a query like ``SELECT %s`` and the output
+        will be a query like ``SELECT ?``.
+
+        :type: str
+        :param operation: A Google BigQuery query string.
+
+        :type: Sequence[Any]
+        :param parameters: Sequence of parameter values.
+        """
+        formatted_params = ['?' for _ in parameters]
+
+        try:
+            return operation % tuple(formatted_params)
+        except TypeError as ex:
+            raise exceptions.ProgrammingError(ex)
+
+    def _format_operation_dict(self, operation, parameters):
+        """Formats parameters in operation in the way BigQuery expects.
+
+        The input operation will be a query like ``SELECT %(namedparam)s`` and
+        the output will be a query like ``SELECT @namedparam``.
+
+        :type: str
+        :param operation: A Google BigQuery query string.
+
+        :type: Mapping[str, Any]
+        :param parameters: Dictionary of parameter values.
+        """
+        formatted_params = {}
+        for name in parameters:
+            formatted_params[name] = '@{}'.format(name)
+
+        try:
+            return operation % formatted_params
+        except KeyError as ex:
+            raise exceptions.ProgrammingError(ex)
+
+    def _format_operation(self, operation, parameters=None):
+        """Formats parameters in operation in way BigQuery expects.
+
+        Raises a ProgrammingError if a parameter used in the operation is not
+        found in the parameters dictionary.
+
+        :type: str
+        :param operation: A Google BigQuery query string.
+
+        :type: Mapping[str, Any] or Sequence[Any]
+        :param parameters: Optional parameter values.
+        """
+        if parameters is None:
+            return operation
+
+        if isinstance(parameters, collections.Mapping):
+            return self._format_operation_dict(operation, parameters)
+
+        return self._format_operation_list(operation, parameters)
+
+    def execute(self, operation, parameters=None):
+        """Prepare and execute a database operation.
+
+        :type: str
+        :param operation: A Google BigQuery query string.
+
+        :type: Mapping[str, Any] or Sequence[Any]
+        :param parameters: Optional dictionary or sequence of parameter values.
+        """
+        self._query_results = None
+        self._page_token = None
+        self._has_fetched_all_rows = False
```

Comment:
1. This is like a `reset()`?
1. This means you expect a single `Cursor` instance to be `execute()`-d many times?
1. Worth having a `reset() / _reset()` helper?